### DISCUSSION TITLE
open ai key

### DISCUSSION THREAD

**Discussion by MohamedALMAHMOUD** [Giskard-AI/giskard-oss#1906] [Q&A] [answered]
Hello,
I try to test your model project . I replaced the os.environ["OPENAI_API_KEY"] by mine. But when I executed the model with langchain, I had this error:

RateLimitError                            Traceback (most recent call last)
[<ipython-input-6-15994558f364>](https://localhost:8080/#) in <cell line: 10>()
      8 text_splitter = RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=100, add_start_index=True)
      9 loader = PyPDFLoader("https://www.ipcc.ch/report/ar6/syr/downloads/report/IPCC_AR6_SYR_LongerReport.pdf")
---> 10 db = FAISS.from_documents(loader.load_and_split(text_splitter), OpenAIEmbeddings())
     11 
     12 # Prepare QA chain

17 frames
[/usr/local/lib/python3.10/dist-packages/openai/api_requestor.py](https://localhost:8080/#) in _interpret_response_line(self, rbody, rcode, rheaders, stream)
    773         stream_error = stream and "error" in resp.data
    774         if stream_error or not 200 <= rcode < 300:
--> 775             raise self.handle_error_response(
    776                 rbody, rcode, resp.data, rheaders, stream_error=stream_error
    777             )

RateLimitError: You exceeded your current quota, please check your plan and billing details. For more information on this error, read the docs: https://platform.openai.com/docs/guides/error-codes/api-errors.
Do you have a solution for testing your model?
Thanks

**Comment by alexcombessie** [comment]
Hey @MohamedALMAHMOUD,

Welcome to the Giskard community! Your error looks like something raised in our Discord community last week, coming from the LLM side. 

@kevinmessiaen would you have an idea of the root cause / fix ?

**Comment by kevinmessiaen** [comment]
Hello @MohamedALMAHMOUD 

It seems that the error is due to you missing credits on openai, you should check your billing plans on openai.

**Reply by MohamedALMAHMOUD** [reply]
Thank you @kevinmessiaen 
So, I can't test your model if I don't have a ChatGpt Plus!

**Reply by kevinmessiaen** [reply]
Sorry I thought that you wanted to know how to use giskard with OpenAI. 

You can test your model using other models providers than OpenAI. We already support `AzureOpenAI`, `Mistral, `Ollama` and you can definitely write your custom model.

You can take a look in the documentation to see example, there is tabs for different model providers: https://docs.giskard.ai/en/latest/open_source/scan/scan_llm/index.html

Let me know if you have trouble to switch to a model that you use.

**Reply by MohamedALMAHMOUD** [reply]
I was testing giskard with OpenAI (but I don't have ChatGpt plus), so I have this error. If you have a solution to use giskard without ChatGpt plus, I'll be happy.

**Reply by kevinmessiaen** [reply]
Since it's the API it should use the pay as you go service and not the ChatGPT plus one: https://openai.com/pricing#language-models

From the FAQ:

> Is API access included in the ChatGPT Plus subscription?
No, OpenAI APIs and ChatGPT Plus subscriptions are billed separately. The API has its own pricing, which can be found at https://openai.com/pricing. The ChatGPT Plus subscription covers usage on chat.openai.com only and costs $20/month.

**Reply by MohamedALMAHMOUD** [reply]
Okay, I understand. So, I can't use giskard, if I don't pay API service. In this case, giskard is no longer open source, is it?

**Reply by kevinmessiaen** [reply, marked as answer]
It's still open source, all our code is accessible on our repository. We are just providing a scan and tests that require you to call LLM models of your choice.

In that particular case we allow you to use the scan with any model and not just OpenAI. Regarding OpenAI, the fact is that you cannot use their API without paying for it and the model are closed source.

But we provide access to open source models such as Mistral and Claude. And if you are running a model by yourself you can define your own `LLMClient` in giskard and call this model by yourself. But it'll probably be way more expensive/slower to run models by yourself rather than using paid services of those models.

If you need helps on how to call others model than OpenAI I can assist you with that.
